src/mongocrypt-kms-ctx.c

```diff
@@ -995,6 +1091,30 @@ static bool _ctx_done_kmip_decrypt(mongocrypt_kms_ctx_t *kms_ctx) {
     return ret;
 }
 
+bool mongocrypt_kms_ctx_fail(mongocrypt_kms_ctx_t *kms) {
```

Comment:
The proposed retry protocol requires retrying network errors immediately after calling `mongocrypt_kms_ctx_fail`. In pseudo-code:

```python
for kms_ctx in mongocrypt_ctx_next_kms_ctx (ctx):
    while True: # Loop to retry network errors.
        try:
            do_socket_operation (kms_ctx)
        except NetworkException as ne:
            if mongocrypt_kms_ctx_fail (kms_ctx):
                continue # Retry request immediately
```

Retryable HTTP errors require less special handling, as they are returned again in `mongocrypt_ctx_next_kms_ctx`.

Suggest changing `mongocrypt_kms_ctx_fail` to set `should_retry` to true. That may help simplify the driver integration. `mongocrypt_ctx_next_kms_ctx` will return the retryable KMS context, regardless of network or HTTP retry. 

Along with the suggestion for an allow-list, and having `mongocrypt_kms_ctx_fail` only return `false` on non-retryable error:

```c
bool mongocrypt_kms_ctx_fail(mongocrypt_kms_ctx_t *kms) {
    if (!kms) {
        return false;
    }

    mongocrypt_status_t *status = kms->status;

    if (!kms->retry_enabled) {
        CLIENT_ERR("KMS request failed due to network error");
        return false;
    }

    if (kms->attempts >= kms_max_attempts) {
        CLIENT_ERR("KMS request failed after %d retries due to a network error", kms_max_attempts);
        return false;
    }

    // Check if request type is retryable. Some requests are non-idempotent and cannot be safely retried.
    _kms_request_type_t retryable_types[] = {MONGOCRYPT_KMS_AWS_ENCRYPT,
                                             MONGOCRYPT_KMS_AWS_DECRYPT,
                                             MONGOCRYPT_KMS_AZURE_WRAPKEY,
                                             MONGOCRYPT_KMS_AZURE_UNWRAPKEY,
                                             MONGOCRYPT_KMS_GCP_ENCRYPT,
                                             MONGOCRYPT_KMS_GCP_DECRYPT};
    bool is_retryable = false;
    for (size_t i = 0; i < sizeof(retryable_types) / sizeof(retryable_types[0]); i++) {
        if (retryable_types[i] == kms->req_type) {
            is_retryable = true;
            break;
        }
    }
    if (!is_retryable) {
        CLIENT_ERR("KMS request failed due to network error");
        return false;
    }

    // Mark KMS context as retryable. Return again in `mongocrypt_ctx_next_kms_ctx`.
    set_retry(kms);

    // Reset intermediate state of parser.
    if (kms->parser) {
        kms_response_parser_reset(kms->parser);
    }
    return true;
}
```